eppo_core/src/background/runtime.rs

```diff
@@ -0,0 +1,218 @@
+    /// need this occasionally (e.g., finish writes to disk, etc.)
+    watched_tasks: TaskTracker,
+}
+
+impl BackgroundRuntime {
+    /// Creates a new `BackgroundRuntime` that runs on the given tokio `runtime`.
+    ///
+    /// The background runtime is active until `stop()` is called.
+    #[must_use]
+    pub fn new(runtime: tokio::runtime::Handle) -> BackgroundRuntime {
+        let cancellation_token = CancellationToken::new();
+        let watched_tasks = TaskTracker::new();
+        BackgroundRuntime {
+            tokio_runtime: runtime,
+            cancellation_token,
+            watched_tasks,
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn tokio_handle(&self) -> &tokio::runtime::Handle {
+        &self.tokio_runtime
+    }
+
+    /// Returns a cancellation token that get cancelled when background runtime is going to
+    /// shutdown.
+    ///
+    /// If you want your task to react to shutdown and perform some cleanup, spawn your task with
+    /// [`BackgroundThread::spawn_tracked()`] and watch this cancellation token.
+    ///
+    /// Note: this is a child token of the main cancellation token, so canceling it does not cancel
+    /// the whole runtime.
+    #[must_use]
+    pub(crate) fn cancellation_token(&self) -> CancellationToken {
+        self.cancellation_token.child_token()
+    }
+
+    pub fn block_on<F: Future>(&self, future: F) -> F::Output {
+        self.tokio_runtime.block_on(future)
+    }
+
+    /// Spawn a task that needs to perform some cleanup on shutdown.
+    ///
+    /// Most tasks shouldn't need that as Rust futures are usually safe to drop.
+    ///
+    /// The task must monitor [`BackgroundRuntime::cancellation_token()`] and exit when the token is
+    /// cancelled.
+    pub(crate) fn spawn_tracked<F>(&self, future: F) -> JoinHandle<F::Output>
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        self.tokio_runtime
+            .spawn(self.watched_tasks.track_future(future))
+    }
+
+    /// Spawn a task that doesn't have any special shutdown requirements.
+    ///
+    /// When runtime is going to shutdown, this task will not be awaited and will be abandoned.
+    ///
+    /// If it's not OK to abandon the task, consider using `spawn_tracked()` instead.
+    pub(crate) fn spawn_untracked<F>(&self, future: F) -> JoinHandle<F::Output>
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        self.tokio_runtime.spawn(future)
+    }
+
+    /// Command background activities to stop and exit.
+    pub(crate) fn stop(&self) {
+        self.watched_tasks.close();
+        self.cancellation_token.cancel();
+    }
+
+    /// Wait for all background activities to stop.
+    pub(super) fn wait(&self) -> impl Future {
+        let tracker = self.watched_tasks.clone();
+        async move { tracker.wait().await }
+    }
```

Comment:
This is intended for background runtime *runner* (i.e., background thread) to know when we're done and it's OK stop tokio runtime.